mwapi/async_session.py

```diff
@@ -0,0 +1,213 @@
+                             .format(prefix))
+
+                if 'error' in doc:
+                    raise APIError.from_doc(doc['error'])
+
+                if 'warnings' in doc:
+                    logger.warning("The following query raised warnings: {0}"
+                                .format(params or data))
+                    for module, warning in doc['warnings'].items():
+                        logger.warning("\t- {0} -- {1}"
+                                    .format(module, warning))
+                return doc
+
+        except (aiohttp.ServerTimeoutError,
+                asyncio.exceptions.TimeoutError) as e:
+            raise TimeoutError(str(e)) from e
+        except aiohttp.ClientConnectionError as e:
+            raise ConnectionError(str(e)) from e
+        except aiohttp.TooManyRedirects as e:
+            raise TooManyRedirectsError(str(e)) from e
+        except Exception as e:
+            raise RequestError(str(e)) from e
+
+
+    async def request(self, method, params=None, query_continue=None,
+                auth=None, continuation=False):
+        """
+        Sends an HTTP request to the API.
+
+        :Parameters:
+            method : `str`
+                Which HTTP method to use for the request?
+                (Usually "POST" or "GET")
+            params : `dict`
+                A set of parameters to send with the request.  These parameters
+                will be included in the POST body for post requests or a query
+                string otherwise.
+            query_continue : `dict`
+                A 'continue' field from a past request.  This field represents
+                the point from which a query should be continued.
+            auth : mixed
+                Auth tuple or callable to enable Basic/Digest/Custom HTTP Auth.
+            continuation : `bool`
+                If true, a continuation will be attempted and a generator of
+                JSON response documents will be returned.
+
+        :Returns:
+            A response JSON documents (or a generator of documents if
+            `continuation == True`)
+        """
+        normal_params = _normalize_params(params, query_continue)
+        if continuation:
+            return self._continuation(method, params=normal_params, auth=auth)
+        else:
+            return await self._request(method, params=normal_params, auth=auth)
+
+    async def _continuation(self, method, params=None, auth=None):
+        if "continue" not in params:
+            params["continue"] = ""
+
+        while True:
+            doc = await self._request(method, params=params, auth=auth)
+            yield doc
+            if "continue" not in doc:
+                break
+            # re-send all continue values in the next call
+            params.update(doc["continue"])
+
+    async def get(self, query_continue=None, auth=None, continuation=False,
+            **params):
```

Comment:
Nit: it would be nicer to read if **params was aligned with self.

Reply:
Done